File: 장세훈/LeetCode/medium/Combinations.java
import java.util.*;

class Combinations {
    /*
        77. Combinations
        https://leetcode.com/problems/combinations/description/
     */

    List<List<Integer>> answer = new ArrayList<>();

    public List<List<Integer>> combine(int n, int k) {
        dfs(n, k, new ArrayList<>(), 1);
        return answer;
    }

    void dfs(int n, int k, List<Integer> list, int start) {
        if (list.size() == k) {
            answer.add(List.copyOf(list));
            return;
        }

        for (int i = start; i <= n; i++) {
            list.add(i);
            dfs(n, k, list, i + 1);
            list.removeLast();
        }
    }
}

File: 정수빈/LeetCode/medium/combinations.py
# https://leetcode.com/problems/combinations
# 설명: 백트레킹, 조합을 구하는 문제 
# 난이도: medium
# 시간: 답지보면서 32분..

class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        n = list(range(1, n+1))
        ans = []
        
        def bk(idx, num):
            if len(num) == k:
                ans.append(num)
                return 
            
            for i in range(idx, len(n)):
                c = n[i]
                bk(i+1, num+[c])
        bk(0, [])
        return ans